### PR TITLE
Implemented several infinite families of named graphs

### PR DESCRIPTION
**Pull request by bspiers**
Implemented some new graph families and added 'wrappers' for some existing named graphs to preserve backwards compatibility.

### CODE REVIEW

File: gap/examples.gi [hunk 1]
```diff
@@ -370,3 +370,488 @@ GeneralisedPetersenGraphCons);
...
+end);
+
+InstallGlobalFunction("DIGRAPHS_BinomialTreeGraph",
+function(n)
+  local D, D_n, bits, i, is2n;
+  if (not IsInt(n)) or n < 1 then
+    ErrorNoReturn("the argument <n> must be an integer greater than 0,");
+    return fail;
+  fi;
+  bits := Log(n, 2);
```
Comment (alex-konovalov) [inline gap/examples.gi:837]: I was looking at what's faster, here there are some timings:
```
gap> for n in [1..100000000] do t:=IsEvenInt(n) and IsPrimePowerInt(n);od;time;
15225
gap> for n in [1..100000000] do bits:=Log(n,2);is2n := n = 2 ^ bits;od;time;
19065
```

Comment (bspiers) [reply gap/examples.gi:837]: Thank you, replaced existing check with faster one

File: gap/examples.gi [hunk 2]
```diff
@@ -370,3 +370,889 @@ GeneralisedPetersenGraphCons);
...
+[IsMutableDigraph, IsInt],
+function(filt, n)
+  local D;
+  if (not IsInt(n)) or n < 1 then
+    ErrorNoReturn("the argument <n> must be an integer greater than 0,");
+  fi;
+  D := Digraph(IsMutableDigraph, []);
+  DigraphAddVertices(D, n);
+  DigraphAddEdges(D, List([1 .. (n - 1)], x -> [x, x + 1]));
+  return DigraphSymmetricClosure(D);
```
Comment (flsmith) [inline gap/examples.gi:1087]: You could just `return DigraphSymmetricClosure(ChainDigraph(IsMutableDigraph, n));`

File: gap/examples.gi [hunk 3]
```diff
@@ -370,3 +370,889 @@ GeneralisedPetersenGraphCons);
...
+InstallMethod(HanoiGraphCons, "for IsMutableDigraph and an integer",
+[IsMutableDigraph, IsInt],
+function(filt, n)
+  local D, C, i, nrVert, exp;
+  if (not IsInt(n)) or n < 1 then
+    ErrorNoReturn("the argument <n> must be an integer greater than 0,");
+  fi;
+  C := CycleGraph(IsMutableDigraph, 3);
+  D := C;
+  for i in [2 .. n] do
```
Comment (flsmith) [inline gap/examples.gi:1179]: Rather than repeatedly creating new digraphs and copying edges over, it would probably be better if you define a single graph with the correct number of nodes and then iteratively construct the graph (using essentially the same method in the loop, something like:

```
DigraphAddEdges(D, Concatenation(DigraphEdges(D) + x, DigraphEdges(D) + (2 * x)));
```

File: gap/examples.gi [hunk 4]
```diff
@@ -370,3 +370,889 @@ GeneralisedPetersenGraphCons);
...
+  fi;
+  bits := Log(n, 2);
+  is2n := IsEvenInt(n) and IsPrimePowerInt(n);
+  if not is2n then
+    bits := bits + 1;
+  fi;
+  D := Digraph(IsMutableDigraph, []);
+  DigraphAddVertex(D, 1);
+  D_n := DigraphMutableCopy(D);
+  for i in [1 .. bits] do
```
Comment (flsmith) [inline gap/examples.gi:1229]: Is it possible to do this without repeatedly copying graphs? I haven't thought about it

**Review by flsmith — CHANGES_REQUESTED** [review]
Lots of good stuff.

After a while I stopped adding comments about `IsPosInt`, but imagine they continue.
All of these new graphs now need documentation (:grimacing:)
In the immutable constructors it would be nice to add any obvious properties that the graphs have.

File: gap/examples.gi [hunk 5]
```diff
@@ -370,3 +370,891 @@ GeneralisedPetersenGraphCons);
...
+
+InstallMethod(WebGraphCons, "for IsMutableDigraph and an integer",
+[IsMutableDigraph, IsInt],
+function(filt, n)
+  local D, i;
+  if (not IsInt(n)) or n < 3 then
+    ErrorNoReturn("the argument <n> must be an integer greater than 2,");
+  fi;
+  D := StackedPrismGraph(IsMutableDigraph, n, 3);
+  for i in [1 .. (n-1)] do
```
Comment (wilfwilson) [inline gap/examples.gi:1144]: This should make the linter happy
```suggestion
  for i in [1 .. (n - 1)] do
```

File: gap/examples.gi [hunk 6]
```diff
@@ -370,3 +370,889 @@ GeneralisedPetersenGraphCons);
...
+InstallMethod(HypercubeGraph, "for a function and an integer",
+[IsFunction, IsInt], HypercubeGraphCons);
+
+InstallMethod(CycleGraphCons, "for IsMutableDigraph and an integer",
+[IsMutableDigraph, IsInt],
+function(filt, n)
+  if (not IsInt(n)) or n < 3 then
+    ErrorNoReturn("the argument <n> must be an integer greater than 2,");
+  fi;
+  return DigraphSymmetricClosure(DigraphMutableCopyIfImmutable(CycleDigraph(n)));
```
[the remark this answers is not in the document]
Comment (wilfwilson) [reply gap/examples.gi:578]: Agreed, Finn's suggestion is better:
```suggestion
  return DigraphSymmetricClosure(CycleDigraph(IsMutableDigraph, n));
```

**Comment by wilfwilson** [comment]
What is the intention for how (if at all) these new things will be documented?

**Comment by bspiers** [comment]
> What is the intention for how (if at all) these new things will be documented?

I'm following the same format as that used in the pancake graph pull request.

File: gap/examples.gi (hunk 4 above)
Comment (bspiers) [reply gap/examples.gi:1229]: > Is it possible to do this without repeatedly copying graphs? I haven't thought about it

There is a definition with labelling vertices by binary numbers and having the subnodes of a node (as a tree) be obtained by changing a lower 0 to a 1, I need to do a test to see which is faster

File: gap/examples.gi (hunk 3 above)
Comment (bspiers) [reply gap/examples.gi:1179]: > Rather than repeatedly creating new digraphs and copying edges over, it would probably be better if you define a single graph with the correct number of nodes and then iteratively construct the graph (using essentially the same method in the loop, something like:
> 
> ```
> DigraphAddEdges(D, Concatenation(DigraphEdges(D) + x, DigraphEdges(D) + (2 * x)));
> ```

I will get on this, I think you're right.

File: doc/examples.xml [hunk 7]
```diff
@@ -743,3 +1393,193 @@ true
...
+
+<#GAPDoc Label="PathGraph">
+<ManSection>
+  <Oper Name="PathGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is a non-negative integer then this operation returns the
+    <M>n</M>th <E>Path graph</E>, consisting of the path on <M>n</M> vertices.
+    This is similar to the Chain digraph but with symmetric edges. The Path 
+    graph has <M>n</M> vertices and <M>n-1</M> edges. The graph is acyclic.<P/>
```
Comment (wilfwilson) [inline doc/examples.xml:1567]: Incorrect, non-empty symmetric digraphs are not acyclic. In particular, for any edge x→y, x→y→x is a cycle.

File: gap/examples.gi [hunk 8]
```diff
@@ -425,6 +425,141 @@ function(filt, n, k)
...
+  vertices := List([1 .. (2 ^ (n - 1))], x -> []);
+  j := 1;
+  for i in [1 .. Length(tuples)] do
+    if Sum(tuples[i]) mod 2 = 0 then
+      vertices[j] := tuples[i];
+      j := j + 1;
+    fi;
+  od;
+  D := Digraph(IsMutableDigraph, []);
+  DigraphAddVertices(D, [1 .. Length(vertices)]);
```
Comment (wilfwilson) [inline gap/examples.gi:550]: Can just do
```suggestion
  D := EmptyDigraph(IsMutableDigraph, Length(vertices));
```

File: gap/examples.gi [hunk 9]
```diff
@@ -502,6 +637,294 @@ function(filt, n, k)
...
+InstallMethod(HypercubeGraphCons, "for IsMutableDigraph and an integer",
+[IsMutableDigraph, IsInt],
+function(filt, n)
+  local D, vertices, i, j;
+  if n < 0 then
+    ErrorNoReturn("the argument <n> must be a non-negative integer,");
+  fi;
+  vertices := Tuples([0, 1], n);
+  D := Digraph(IsMutableDigraph, []);
+  DigraphAddVertices(D, [1 .. Length(vertices)]);
```
Comment (wilfwilson) [inline gap/examples.gi:669]:
```suggestion
  D := EmptyDigraph(IsMutableDigraph, Length(vertices));
```

File: gap/examples.gi [hunk 10]
```diff
@@ -425,6 +425,141 @@ function(filt, n, k)
...
+          [1, -1]];
+  H_n := [[1]];
+  if n > 1 then
+    for i in [2 .. n] do
+      H_n := KroneckerProduct(H_2, H_n);
+    od;
+  fi;
+  sideHn := Length(H_n);
+  D := Digraph(IsMutableDigraph, []);
+  DigraphAddVertices(D, 4 * sideHn);
```
Comment (wilfwilson) [inline gap/examples.gi:503]:
```suggestion
  D := EmptyDigraph(IsMutableDigraph, 4 * sideHn);
```

File: gap/examples.gi [hunk 11]
```diff
@@ -370,3 +370,889 @@ GeneralisedPetersenGraphCons);
...
+
+InstallMethod(WheelGraph, "for a function and an integer",
+[IsFunction, IsInt], WheelGraphCons);
+
+InstallMethod(HelmGraphCons, "for IsMutableDigraph and an integer",
+[IsMutableDigraph, IsInt],
+function(filt, n)
+  local D;
+  if (not IsInt(n)) or n < 3 then
+    ErrorNoReturn("the argument <n> must be an integer greater than 2,");
```
[the remark this answers is not in the document]
Comment (wilfwilson) [reply gap/examples.gi:772]: No, the method creates the wheel graph with `n+1` vertices.

File: gap/examples.gi [hunk 12]
```diff
@@ -658,6 +1081,307 @@ function(filt, m, n)
...
+  D := MakeImmutable(KneserGraphCons(IsMutableDigraph, n, k));
+  SetIsMultiDigraph(D, false);
+  SetIsSymmetricDigraph(D, true);
+  SetIsRegularDigraph(D, true);
+  SetIsVertexTransitive(D, true);
+  SetIsEdgeTransitive(D, true);
+  if n >= 2 * k then
+    SetChromaticNumber(D, n - 2 * k + 2);
+  else
+    SetChromaticNumber(D, 1);
```
Comment (wilfwilson) [inline gap/examples.gi:1098]: Could you please add a test for this line? Also, the only digraphs with chromatic number 1 are empty, so I presume you could/should also set that property too, as you did for e.g. `KellerGraphCons`?
```suggestion
    SetChromaticNumber(D, 1);
    SetIsEmptyDigraph(D, true);
```

File: gap/examples.gi [hunk 13]
```diff
@@ -711,6 +1435,110 @@ function(filt, m, n)
...
+  SetIsAcyclicDigraph(D, true);
+  SetIsEmptyDigraph(D, n = 1);
+  return D;
+end);
+
+InstallMethod(PathGraph, "for an integer", [IsPosInt],
+n -> PathGraphCons(IsImmutableDigraph, n));
+
+InstallMethod(PathGraph, "for a function and an integer",
+[IsFunction, IsPosInt], PathGraphCons);
```
Comment (wilfwilson) [inline gap/examples.gi:1484]: @LRacine has also implemented the same graph in #457 - I guess only one implementation is needed. You can decide amongst yourselves which one to keep!

File: gap/examples.gi (hunk 2 above)
Comment (wilfwilson) [reply gap/examples.gi:1464]: Agreed! In any case, even if `ChainDigraph` didn't exist, it's not worth it to create an empty digraph, add the necessary number of vertices, and then add the edges all in separate steps - you could just directly do:
```
D := Digraph(IsMutableDigraph, List([1 .. (n - 1)], x -> [x, x + 1]));
```

**Review by wilfwilson — CHANGES_REQUESTED** [review]
Thanks for doing all o f this! I have recently squashed and rebased all of your commits, and resolved the many (many!) merge conflicts in the process. It should therefore now be easy to make further changes to this, and to see what is failing on the tests.

(I recommend in the future to do a larger number of smaller PRs; each one is quick and easier to create, review, and get merged, and there is a smaller scope for merge conflicts.)

Since this PR is very large, I haven't given it a completely thorough review, but I note that @flsmith has also given this a bit of a review, and perhaps other people are going to, as well. But I've given it a bit of a skim and added some specific comments.

Unfortunately, if other PRs are merged before this one, it's likely that there will be more merge conflicts. If you need help resolving these, just let one of us know. (Note that my resolution of the previous conflicts evidently didn't go perfectly, since now there are some methods which are no longer adjacent in the file, even though they presumably used to be).

File: gap/examples.gi [hunk 14]
```diff
@@ -658,6 +1081,307 @@ function(filt, m, n)
...
+n -> OddGraphCons(IsImmutableDigraph, n));
+
+InstallMethod(OddGraph, "for a function and an integer",
+[IsFunction, IsInt], OddGraphCons);
+
+InstallMethod(KellerGraphCons, "for IsMutableDigraph and an integer",
+[IsMutableDigraph, IsInt],
+function(filt, n)
+  local D, vertices, i, j;
+  if (not IsInt(n)) or n < 0 then
```
Comment (wilfwilson) [inline gap/examples.gi:1184]: This is more of the sort of thing that @flsmith was talking about - `n` is already known to be an integer by this point.
```suggestion
  if n < 0 then
```

File: gap/examples.gi [hunk 15]
```diff
@@ -370,3 +370,889 @@ GeneralisedPetersenGraphCons);
...
+  return D;
+end);
+
+InstallMethod(LindgrenSousselierGraph, "for an integer", [IsInt],
+n -> LindgrenSousselierGraphCons(IsImmutableDigraph, n));
+
+InstallMethod(LindgrenSousselierGraph, "for a function and an integer",
+[IsFunction, IsInt], LindgrenSousselierGraphCons);
+
+InstallMethod(HypohamiltonianGraphCons, "for IsMutableDigraph and an integer",
```
[the remark this answers is not in the document]
Comment (bspiers) [reply gap/examples.gi]: Renamed to "BondyGraph" to be more clear

File: doc/examples.xml [hunk 16]
```diff
@@ -408,6 +445,85 @@ gap> OutNeighbors(D);
...
+
+<#GAPDoc Label="MobiusLadderGraph">
+<ManSection>
+  <Oper Name="MobiusLadderGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is a positive integer at least 4, then this operation returns
+    the <E>Mobius Ladder graph</E> that is obtained by introducing a 'twist'
+    in the <M>n</M>th Prism graph, similar to the construction of a Mobius
+    strip. The Mobius ladder graph is isomorphic to the Circulant graph
```
Comment (james-d-mitchell) [inline doc/examples.xml:494]:
```suggestion
    strip. The Mobius ladder graph is isomorphic to the circulant graph
```

**Comment by wilfwilson** [comment]
Thanks for your further work on this @bspiers, sorry for the further delay in reviewing.

I've now squashed and rebased and resolved merge conflicts, and organised the new contributions into (internal) alphabetical order. Let's see if the tests pass now.

I haven't looked at the implementations in detail, but I'll give the code a quick look again soon (but on the whole your implementations certainly seem reasonable).

File: doc/examples.xml [hunk 17]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+<#GAPDoc Label="AndrasfaiGraph">
+<ManSection>
+  <Oper Name="AndrasfaiGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is an integer greater than 0, then this operation returns the
+    <M>n</M>th <E>Andrasfai graph</E>. The Andrasfai graph is a circulant graph
+    with <M>3n - 1</M> vertices. The indices of the Andrasfai graph are given by
+    the numbers between <M>1</M> and <M>3n - 1</M> that are congruent to
+    <M>1 mod 3</M> (that is, for each index <M>j</M>, vertex <M>i</M> is
```
Comment (wilfwilson) [inline doc/examples.xml:869]:
```suggestion
    the numbers between <C>1</C> and <C>3<A>n</A> - 1</C> that are congruent to
    <C>1</C> mod <C>3</C> (that is, for each index <M>j</M>, vertex <M>i</M> is
```

File: doc/examples.xml [hunk 18]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+    adjacent to the <M>i + j</M>th and <M>i - j</M> vertices). The graph has
+    <M>6(3n - 1)</M> edges. The graph is triangle free.<P/>
```
Comment (wilfwilson) [inline doc/examples.xml:871]:
```suggestion
    <C>6(3<A>n</A> - 1)</C> edges. The graph is triangle free.<P/>
```

File: doc/examples.xml [hunk 19]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is a positive integer then this operation returns the
+    <M>n</M>th <E>binomial tree graph</E>. The binomial tree graph has <M>n</M>
+    vertices and <M>n-1</M> undirected edges. The vertices of the binomial tree
+    graph are the numbers from 1 to <M>n</M> in binary representation, with a
+    vertex <M>v</M> having as a direct parent the vertex with binary
+    representation the same as <M>v</M> but with the lowest 1-bit cleared. For
+    example, the vertex <M>011</M> has parent <M>010</M>, and the vertex
+    <M>010</M> has parent <M>000</M>.<P/>
```
Comment (wilfwilson) [inline doc/examples.xml:905]:
```suggestion
    If <A>n</A> is a positive integer then this operation returns the
    <A>n</A>th <E>binomial tree graph</E>. The binomial tree graph has <A>n</A>
    vertices and <C><A>n</A>-1</C> undirected edges. The vertices of the binomial tree
    graph are the numbers from 1 to <A>n</A> in binary representation, with a
    vertex <C>v</C> having as a direct parent the vertex with binary
    representation the same as <C>v</C> but with the lowest 1-bit cleared. For
    example, the vertex <M>011</M> has parent <M>010</M>, and the vertex
    <M>010</M> has parent <M>000</M>.<P/>
```

File: doc/examples.xml [hunk 20]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+<ManSection>
+  <Oper Name="BondyGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is a non-negative integer then this operation returns the
+    <M>n</M>th <E>Bondy graph</E>. The Bondy graphs are a family of
+    hypohamiltonian graphs: a graph which is not Hamiltonian itself but the
+    removal of any vertex produces a Hamiltonian graph. The Bondy graphs are
+    the <M>(3 * (2 * n + 1) + 2, 2)</M>th Generalised Petersen graphs, and have
+    <M>12n + 10</M> vertices and <M>15 + 18n</M> undirected edges.<P/>
```
Comment (wilfwilson) [inline doc/examples.xml:932]:
```suggestion
    If <A>n</A> is a non-negative integer then this operation returns the
    <C>n</C>th <E>Bondy graph</E>. The Bondy graphs are a family of
    hypohamiltonian graphs: a graph which is not Hamiltonian itself but the
    removal of any vertex produces a Hamiltonian graph. The Bondy graphs are
    the <C>(3 * (2 * <A>n</A> + 1) + 2, 2)</C>th Generalised Petersen graphs, and have
    <C>12<A>n</A> + 10</C> vertices and <C>15 + 18<A>n</A></C> undirected edges.<P/>
```

File: doc/examples.xml [hunk 21]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+    a <E>circulant graph</E>. The circulant graph is a graph on <A>n</A>
+    vertices, where for each element <M>j</M> of <A>par</A>, the <M>i</M>th
+    vertex is adjacent to the <M>(i + j)</M>th and <M>(i - j)</M>th
+    vertices.<P/>
+
+    If <A>par</A> is <M>[1]</M>, then the graph is the <M>n</M>th cyclic graph.
+    If <A>par</A> is <M>[1, 2, . . ., Int(n / 2)]</M>, then the graph is the
+    complete graph on <A>n</A> vertices. If <A>n</A> is at least 4 and
+    <A>par</A> is <M>[1, n]</M> then the graph is the <M>n</M>th Mobius ladder
+    graph.<P/>
```
Comment (wilfwilson) [inline doc/examples.xml:967]:
```suggestion
    If <A>n</A> is an integer greater than 1, and <A>par</A> is a list of
    integers that are contained in <C>[1..<A>n</A>]</C> then this operation returns
    a <E>circulant graph</E>. The circulant graph is a graph on <A>n</A>
    vertices, where for each element <M>j</M> of <A>par</A>, the <M>i</M>th
    vertex is adjacent to the <M>(i + j)</M>th and <M>(i - j)</M>th
    vertices.<P/>

    If <A>par</A> is <M>[1]</M>, then the graph is the <A>n</A>th cyclic graph.
    If <A>par</A> is <C>[1,2,...,Int(<A>n</A>/2)]</C>, then the graph is the
    complete graph on <A>n</A> vertices. If <A>n</A> is at least 4 and
    <A>par</A> is <C>[1,<A>n</A>]</C> then the graph is the <A>n</A>th Mobius ladder
    graph.<P/>
```

File: doc/examples.xml [hunk 22]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+<ManSection>
+  <Oper Name="CycleGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is an integer greater than 2 then this operation returns the
+    <M>n</M>th <E>cycle graph</E>, consisting of the cycle on <M>n</M> vertices.
+    The cycle graph, unlike the cycle digraph, is symmetric. The cycle graph
+    has <M>n</M> vertices and <M>n</M> undirected edges. The cycle graph is
+    simple so the non-simple graphs with a single vertex and single loop and
+    with two vertices and two edges between them are excluded.<P/>
```
Comment (wilfwilson) [inline doc/examples.xml:997]:
```suggestion
    If <A>n</A> is an integer greater than 2 then this operation returns the
    <A>n</A>th <E>cycle graph</E>, consisting of the cycle on <A>n</A> vertices.
    The cycle graph, unlike the cycle digraph, is symmetric. The cycle graph
    has <A>n</A> vertices and <A>n</A> undirected edges. The cycle graph is
    simple so the non-simple graphs with a single vertex and single loop and
    with two vertices and two edges between them are excluded.<P/>
```

File: doc/examples.xml [hunk 23]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+  <Description>
+    If <A>n</A> is a positive integer at least 3, then this operation
+    returns the <E>gear graph</E> with <M>2n + 1</M> vertices and
+    <M>3n</M> undirected edges. The <M>n</M>th gear graph is the <M>2n</M>th
+    cycle graph with one additional central vertex, to which every other
+    vertex of the cycle is connected. The gear graph is a symmetric digraph.
+    A gear graph is a Matchstick graph, that is it is simple with a planar
+    graph embedding, and is a unit-distance graph (that is, it can be embedded
+    into the Euclidean plane with vertices being distinct points and edges
+    having length 1).<P/>
```
Comment (wilfwilson) [inline doc/examples.xml:1024]:
```suggestion
    If <A>n</A> is a positive integer at least 3, then this operation
    returns the <E>gear graph</E> with <C>2<A>n</A> + 1</C> vertices and
    <C>3<A>n</A></A> undirected edges. The <A>n</A>th gear graph is the <C>2<A></A></C>th
    cycle graph with one additional central vertex, to which every other
    vertex of the cycle is connected. The gear graph is a symmetric digraph.
    A gear graph is a Matchstick graph, that is it is simple with a planar
    graph embedding, and is a unit-distance graph (that is, it can be embedded
    into the Euclidean plane with vertices being distinct points and edges
    having length 1).<P/>
```

File: doc/examples.xml [hunk 24]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+<ManSection>
+  <Oper Name="HalvedCubeGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is a positive integer at least 1, then this operation returns
+    the <M>n</M>th <E>halved cube graph</E>, the graph of the <M>n</M>-
+    demihypercube. The vertices of the graph are those of the <M>n</M>th-
+    hypercube, with two vertices adjacent if and only if they are at distance
+    1 or 2 from each other. Equivalent constructions are as the second graph
+    power of the <M>n-1</M>th hypercube graph, or as with vertices labelled as
```
Comment (wilfwilson) [inline doc/examples.xml:1052]:
```suggestion
    If <A>n</A> is a positive integer at least 1, then this operation returns
    the <A>n</A>th <E>halved cube graph</E>, the graph of the <A>n</A>-
    demihypercube. The vertices of the graph are those of the <A>n</A>th-
    hypercube, with two vertices adjacent if and only if they are at distance
    1 or 2 from each other. Equivalent constructions are as the second graph
    power of the <C><A>n</A>-1</C>th hypercube graph, or as with vertices labelled as
```

File: doc/examples.xml [hunk 25]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+<#GAPDoc Label="HanoiGraph">
+<ManSection>
+  <Oper Name="HanoiGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is positive integer then this operation returns the <M>n</M>th
+    <E>Hanoi graph</E>. The Hanoi graph's vertices represent the possible states
+     of the 'Tower of Hanoi' puzzle on three 'towers', while its edges represent
+     possible moves. The Hanoi graph has <M>3^n</M> vertices, and
+     <M>(1/2) * Binomial(3, 3) * (3^n - 1)</M> undirected edges.<P/>
```
Comment (wilfwilson) [inline doc/examples.xml:1085]:
```suggestion
    If <A>n</A> is positive integer then this operation returns the <A>n</A>th
    <E>Hanoi graph</E>. The Hanoi graph's vertices represent the possible states
     of the 'Tower of Hanoi' puzzle on three 'towers', while its edges represent
     possible moves. The Hanoi graph has <C>3^<A>n</A></C> vertices, and
     <C>Binomial(3, 3) * (3^<A>n</A> - 1) / 2</C> undirected edges.<P/>
```

File: doc/examples.xml [hunk 26]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+</ManSection>
+<#/GAPDoc>
+
+<#GAPDoc Label="HelmGraph">
+<ManSection>
+  <Oper Name="HelmGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is a positive integer at least 3, then this operation returns
+    the <M>n</M>th <E>helm graph</E>. The helm graph is the <A>n</A>-1th wheel
```
Comment (wilfwilson) [inline doc/examples.xml:1114]:
```suggestion
    If <A>n</A> is a positive integer at least 3, then this operation returns
    the <A>n</A>th <E>helm graph</E>. The helm graph is the <A>n</A>-1th wheel
```

File: doc/examples.xml [hunk 27]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+
+<#GAPDoc Label="HypercubeGraph">
+<ManSection>
+  <Oper Name="HypercubeGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is a non-negative integer, then this operation returns the
+    <M>n</M>th <E>hypercube graph</E>. The graph has <M>2^n</M> vertices and
+    <M>2^(n-1)n</M> edges. It is formed from the vertices and edges of the
+    <M>n</M>-dimensional hypercube. Alternatively, the graph can be constructed
```
Comment (wilfwilson) [inline doc/examples.xml:1139]:
```suggestion
    If <A>n</A> is a non-negative integer, then this operation returns the
    <A>n</A>th <E>hypercube graph</E>. The graph has <C>2^<A>n</A></C> vertices and
    <C>2^(<A>n</A>-1)<A>n</A></C> edges. It is formed from the vertices and edges of the
    <A>n</A>-dimensional hypercube. Alternatively, the graph can be constructed
```

File: doc/examples.xml [hunk 28]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+  <Oper Name="KellerGraph" Arg="[filt, ]n"/>
+  <Returns>A digraph.</Returns>
+  <Description>
+    If <A>n</A> is an integer at least 0 then this operation returns the
+    <M>n</M>th or <M>n</M>-dimensional <E>Keller graph</E>. The graph has
+    vertices given by the <M>n</M>-tuples on the set <M>[0, 1, 2, 3]</M>.
+    Two vertices are adjacent if their respective tuples are such that they
+    differ in at least two coordinates and in at least one coordinate the
+    difference between the two is <M>2 mod 4</M>. The Keller graph has
+    <M>4 ^ n</M> vertices.<P/>
```
Comment (wilfwilson) [inline doc/examples.xml:1170]:
```suggestion
    If <A>n</A> is a nonnegative integer then this operation returns the
    <A>n</A>th or <A>n</A>-dimensional <E>Keller graph</E>. The graph has
    vertices given by the <A>n</A>-tuples on the set <M>[0, 1, 2, 3]</M>.
    Two vertices are adjacent if their respective tuples are such that they
    differ in at least two coordinates and in at least one coordinate the
    difference between the two is <M>2 mod 4</M>. The Keller graph has
    <C>4 ^ <A>n</A></C> vertices.<P/>
```

File: doc/examples.xml [hunk 29]
```diff
@@ -856,3 +856,744 @@ gap> BinaryTree(IsMutableDigraph, 8);
...
+
+    The Keller graphs were constructed with the intention of finding
+    counterexamples to Keller's conjecture
+    (<URL>https://mathworld.wolfram.com/KellersConjecture.html</URL>), and has
+    been used since for testing maximum clique algorithms.<P/>
+
+    If <M>n</M> is 1 then the graph is empty, for <M>n</M> greater than 1 the
+    chromatic number of the Keller graph is <M>2^n</M> and the graph is
```
Comment (wilfwilson) [inline doc/examples.xml:1178]:
```suggestion
    If <A>n</A> is 1 then the graph is empty, for <A>n</A> greater than 1 the
    chromatic number of the Keller graph is <C>2^<A>n</A></C> and the graph is
```

**Review by wilfwilson — CHANGES_REQUESTED** [review]
I started making suggestions in the GitHub interface about tagging named arguments in the documentation with `<A>` tags, both when they are freestanding, and when they are in code/maths (which requires using `<C>` tags rather than `<M>` in these cases).

I will do the rest directly in the code, it's much easier.
